src/payjoin/app.rs

```diff
@@ -0,0 +1,214 @@
+        let _to_broadcast_in_failure_case = proposal.extract_tx_to_schedule_broadcast();
+
+        // The network is used for checks later
+        let network = self.app.network();
+
+        // Receive Check 1: Can Broadcast
+        let proposal = proposal.check_broadcast_suitability(None, |tx| {
+            let _raw_tx = bitcoin::consensus::encode::serialize_hex(&tx);
+            // TODO test_mempool_accept e.g.:
+            //
+            // let mempool_results =
+            //    bitcoind.test_mempool_accept(&[raw_tx]).map_err(|e| Error::Server(e.into()))?;
+            // match mempool_results.first() {
+            //     Some(result) => Ok(result.allowed),
+            //     None => Err(Error::Server(
+            //         anyhow!("No mempool results returned on broadcast check").into(),
+            //     )),
+            // }
+            Ok(true)
+        })?;
+        tracing::trace!("check1");
+
+        use bdk::bitcoin::{Script, ScriptBuf};
+        use std::sync::mpsc;
+        use std::sync::mpsc::Sender;
+        use std::sync::Arc;
+        use std::sync::Mutex;
+
+        // Create a channel
+        let (tx, rx) = mpsc::channel::<(ScriptBuf, Sender<Result<bool, ApplicationError>>)>();
+        let tx = Arc::new(Mutex::new(tx));
+
+        let app = self.app.clone();
+        let profile = self.profile.clone();
+        let network = network.clone();
+
+        // Receive Check 2: receiver can't sign for proposal inputs
+        tokio::spawn(async move {
+            while let Ok((input, response_tx)) = rx.recv() {
+                let result = check_script_owned(&app, &profile, network, &input).await;
+                let _ = response_tx.send(result);
+            }
+        });
+
+        async fn check_script_owned(
+            app: &App,
+            profile: &Profile,
+            network: bitcoin::Network,
+            input: &Script,
+        ) -> Result<bool, ApplicationError> {
+            let address = bitcoin::BdkAddress::from_script(&input, network)?;
+            match app
+                .find_address(&profile.clone(), address.to_string())
+                .await
+            {
+                Err(ApplicationError::AddressError(e)) => Err(ApplicationError::AddressError(e)),
+                Err(_) => Ok(false), // good address script, but not found
+                Ok(_) => Ok(true),
+            }
+        }
+
+        let proposal = proposal.check_inputs_not_owned(|input| {
+            let (response_tx, response_rx) = mpsc::channel::<Result<bool, ApplicationError>>();
+            {
+                let tx = tx.lock().expect("lock"); // TODO Handle lock error if needed
+                tx.send((input.to_owned(), response_tx))
+                    .map_err(|e| Error::Server(e.into()))?;
+            }
+            let recv_res = response_rx.recv().map_err(|e| Error::Server(e.into()))?;
+            Ok(recv_res.map_err(|e| Error::Server(e.into()))?)
+        })?;
```

Comment:
This hack looks ugly but is necessary with the current `payjoin` library since `proposal.check_*` functions take closures, async closures are unstable, and the payjoin app needs to make async db calls for various checks if e.g. a particular address is ours or not in this case.

Open to suggestions for improvements here in either implementation approach or upstream api.

Reply:
The most straightforward solution is to block_on a `tokio::runtime::Handle`
Justin and I explored async_trait too which will expand the scope of the library. I'd like to avoid that for now.

```rs
        // Receive Check 2: receiver can't sign for proposal inputs
        let proposal = proposal.check_inputs_not_owned(|input| {
            self.rt.block_on(async {
                let address = bitcoin::BdkAddress::from_script(&input, network)
                    .map_err(|e| Error::Server(e.into()))?;
                match self.addresses.find_address(address.to_string()).await {
                    Ok(_) => Ok(true),
                    Err(AddressError::AddressNotFound(_)) => Ok(false),
                    Err(e) => Err(Error::Server(e.into())),
                }
            })
        })?;
```